### DISCUSSION TITLE
golevelup/nestjs-rabbitmq RabbitMQModule initializing with a cluster of 3 rabbitmq servers

### DISCUSSION THREAD

**Discussion by satishkotturi** [golevelup/nestjs#745] [Q&A]
is there a way to initialize the RabbigMQModule to connect to a cluster of 3 rabbitmq server in round robin if any one of the server goes down.
example below:
RabbitMQModule.forRoot(RabbitMQModule, {
  uri: [`amqp://${config.user}:${config.password}@${config.**hostOne**}:${configuration().amqp.port}`,
`amqp://${config.user}:${config.password}@${config.**hostTwo**}:${configuration().amqp.port}`,
`amqp://${config.user}:${config.password}@${config.**hostThree**}:${configuration().amqp.port}`
  ],
],
});

The above initialization is failing if the hostOne is down.
Note: The same initialization is working if I configured only one server( for the uri param as a string )and it is up and running.

**Comment by BeyondEvil** [comment]
Would like to know this as well.

**Comment by ramioooz** [comment]
Hello @satishkotturi 

Yes — `@golevelup/nestjs-rabbitmq` can work with multiple RabbitMQ node URLs, because it uses `amqp-connection-manager` underneath, and that library supports a list of brokers plus reconnect/failover behavior. RabbitMQ clients still connect to **one node at a time**, and on node failure they should reconnect to another node from the list.  [reference](https://github.com/golevelup/nestjs/blob/master/packages/rabbitmq/README.md)

The original code was failing because:
- When `uri` is passed as an **array**, the library (which uses amqp-connection-manager)
  tries the hosts in order.
- By default `connectionInitOptions.wait` is `true`, so if the first host (hostOne)
  is down during app bootstrap, the whole module initialization throws and the app
  never starts.

The same single-string URI configuration worked fine, which is why the issue only
appeared with the cluster setup.

Fixed it with the officially supported cluster pattern:

```ts
RabbitMQModule.forRoot(RabbitMQModule, {
  uri: [
    `amqp://${config.user}:${config.password}@${config.hostOne}:${configuration().amqp.port}`,
    `amqp://${config.user}:${config.password}@${config.hostTwo}:${configuration().amqp.port}`,
    `amqp://${config.user}:${config.password}@${config.hostThree}:${configuration().amqp.port}`,
  ],

  // This is the key piece that was missing
  connectionInitOptions: {
    wait: false,        // don't block bootstrap if first host is temporarily down
  },

  // ...rest of your existing options (exchanges, prefetchCount, etc.)
});
```
Notes:

* amqp-connection-manager automatically cycles through the hosts on failure → true round-robin failover.
* You can reorder the array if you have a preferred primary node.
* No changes needed on the RabbitMQ side.
* Tested with one node down — connects instantly to the next healthy one.
* Works on the current version of @golevelup/nestjs-rabbitmq.

This should close the discussion. Hope it helps anyone else hitting the same issue!

**Comment by Tanx-1811** [comment]
Yes. You can use multiple RabbitMQ URIs with `connectionInitOptions.wait = false`:

```ts
RabbitMQModule.forRoot(RabbitMQModule, {
  uri: [
    `amqp://${user}:${password}@${hostOne}:5672`,
    `amqp://${user}:${password}@${hostTwo}:5672`,
    `amqp://${user}:${password}@${hostThree}:5672`,
  ],
  connectionInitOptions: {
    wait: false,
  },
});
```

`wait: false` prevents the NestJS application from failing during startup if the first RabbitMQ server is down. The connection manager will continue trying to connect and fail over to another available server.

Note: this provides **failover**, not true round-robin load balancing.
